### ISSUE TITLE
Allow resource owners to set resource 'shareable' flag

### ISSUE THREAD

**Issue by pkdash** [hydroshare/hydroshare#664]
When a resource is created by default the 'shareable' flag is set true  meaning resource can be shared by any users who have some level of access permission to the resource. Only resource owners can set this flag. If 'shareable' is set to false, only resource owners can share the resource with other users.

Pabitra to implement backend support for this functionality
Mauriel to implement frontend support for this functionality

**Comment by mauriel133** [comment]
By shareable do you mean the ability to grant permissions to other users?

**Comment by pkdash** [comment]
Yes. When the resource is not shareable a user who has either 'edit' or 'view' permission on that resource can't give access to that resource to any other users. However, the resource owners can always give permissions to other users.

**Comment by dtarb** [comment]
Following is the definition of the sharing attribute in the Access Control Policy

Shareable – Entities other than owners may grant access to the resource. If False, non-owners who receive access to the resource may not grant that access to others. 

Sharing (extending access)
The boolean attribute ‘Shareable’ is intended to give resource owners control over who the resource may be distributed to, a use case that is deemed important to protect the intellectual content of resources that may contain unpublished data.  Shareable=False effectively holds the rights to enable access to another user at the owner level.  Shareable=True extends these rights to users with change and view privileges, allowing redistribution of the resource at the same level of permissions that a user has, i.e. the resource is "Distributable", even for users with read-only privilege.  Specifically, when Shareable=True then a user with view settings can share the resource at the view level (i.e. enable view access to another user or group in the system).  When Shareable=True, a user with change settings can share the resource at the change or the view level.  Only owners may change the ‘Shareable’ setting.

The full policy document is in Alfresco at: https://my.alfresco.com/share/s/WJuMf2ghRlyuty0oXrqe5g

Please do review the full policy as we will eventually need an interface that allows management of all the access control attributes.

**Comment by pkdash** [comment]
@mauriel133 I have now fixed this issue(both front and backend). Can you make the necessary adjustment to the fontend styling may be? Branch:664-using-resource-shareable-flag. Once you are done, I can create the PR for it.

**Comment by mauriel133** [comment]
Sure, I will have a look.

**Comment by mauriel133** [comment]
This PR should wait for the new access control interface to be merged in order to avoid conflicts and working on it twice. Once the interface is merged I will bring this one up to speed.

It looks to me that this should be a toggle button. When enabled, view users should have access to the access panel when the resource is in view mode (since they are not allowed to go into edit view). This will make things a little inconsistent with the respect to the access control interface, which right now is only accessible through edit mode.

Should we bring the access control interface back into view mode for every user that can grant at least one level of permission?

**Comment by pkdash** [comment]
@mauriel133 There is no PR for it at this point. Once you make the necessary changes to this branch, I can create a PR. If you think this should wait until the new access control interface gets merged then we will wait to create the PR until then.

**Comment by horsburgh** [comment]
@mauriel133 @pkdash and @dtarb - this inconsistency seems complicated to me.  Do we really want the access control panel to show outside of edit mode for users that have view only access? We just moved this functionality to the edit page so that all editing is done consistently in one place. And, do we also want to allow view only users to actually edit the access control for a resource (e.g., to add more view only users)?  I just want to make sure we are clear on the implications of this.

**Comment by dtarb** [comment]
@horsburgh we do want users with view permissions to be able to add new view users for resources for which "shareable" is set.  This is to encourage sharing, and was a policy decision made without considering how the functionality would be implemented.  I recognize that the choice to have access control on the edit page that is only accessible to users with edit permissions makes this complicated, but would prefer that design choices about the layout do not dictate policy.

The reason for the policy is to reduce copying and offline sharing of content, as if a user could not share a resource with someone, they may be tempted to make a copy or download and share it separately.  It is better to keep the collaboration in HydroShare.

**Comment by horsburgh** [comment]
@dtarb - I understand the desires behind the policy.  But, the current implementation would be overly complicated as it is now. We should make policy decisions with some consideration about how the functionality would be implemented.

Since the access control panel needs to be accessible outside of editing mode for view only users, I suggest we move it outside of editing mode for everyone.  If a user has any permissions that would allow them to grant others permission, a button would appear on the resource landing page outside of editing mode allowing them to click to get the new access control panel.  Unfortunately, I think this undoes some work that @mauriel133 has already done, but may not be too involved.

**Comment by dtarb** [comment]
Policy did not dictate that you had to implement on separate pages.  That was the choice that led to the complication.

**Comment by ironfroggy** [comment]
Implementation and maintenance costs should absolutely be an important consideration in policy, both in terms of defining overall goals and in prioritizing what has a greater ROI to deliver sooner rather than later.

**Comment by pkdash** [comment]
@mauriel133 Wondering on the status of UI changes needed for setting the 'shareable' flag.

**Comment by mauriel133** [comment]
I didn't make progress on this issue during the break, but I'm working on it this week.

**Comment by mauriel133** [comment]
Is there anything else that needs to be done for this functionality?

**Comment by dtarb** [comment]
Tested and working to my liking.
